README
NAME
    Class::DBI::Lite - Lightweight ORM for Perl

SYNOPSIS
    Create some database tables:

      create table artists (
        artist_id integer primary key autoincrement,
        artist_name varchar(100) not null
      );
      
  create table cds (
        cd_id integer primary key autoincrement,
        artist_id integer not null,
        cd_name varchar(100) not null
      );

      package My::Model;
      
  use base 'Class::DBI::Lite';
      
  __PACKAGE__->connection( 'DBI:mysql:dbname:localhost', 'user', 'pass' );
      
  1;# return true:

      package My::Artist;
      
  use base 'My::Model';
      
  __PACKAGE__->set_up_table('artists');
      
  __PACKAGE__->has_many(
        cds =>
          'My::CD' =>
            'artist_id'
      );
      
  1;# return true:

      package My::CD;
      
  use base 'My::Model';
      
  __PACKAGE__->set_up_table('cds');
      
  __PACKAGE__->has_a(
        artist =>
          'My::Artist' =>
            'artist_id'
      );
      
  1;# return true:

    Then, in your script someplace:

      use My::Artist;
      
  my $artist = My::Artist->retrieve( 123 );
      
  foreach my $cd ( $artist->cds )
      {
        ...
      }# end foreach()
      
  my $cd = $artist->add_to_cds( cd_name => "Attak" );
      
  print $cd->cd_name;
      $cd->cd_name("New Name");
      $cd->update();
      
  # Delete the artist and all of its CDs:
      $artist->delete;

DESCRIPTION
    Sometimes Class::DBI is too crufty, and DBIx::Class is too much.

    Enter Class::DBI::Lite.

TODO
    *   Complete tests

    *   Examples

    *   Documentation

AUTHOR
    Copyright John Drago <jdrago_999@yahoo.com>. All rights reserved.

LICENSE
    This software is Free software and may be used and redistributed under
    the same terms as perl itself.



Changes
Revision history for Perl extension Class::DBI::Lite.

1.033   2012-03-08
  - This release fixes a long-standing bug in Class::DBI::Lite::Pager->nagivations
    which caused it to sometimes return '0' as the "start" position.
    eg: my ($start, $stop) = $pager->navigations()
    If you then did:
    for my $page_number ($start..$stop) { # Make a hyperlink to $page_number }
    Your first $page_number would sometimes be '0'
  - Ray Baksh++ for submitting the patch at a ridiculous wee hour in the morning.
    (Much appreciated!)

1.032   2012-03-02
  - No longer depend on or use DBIx::ContextualFetch.
  - RootMeta was not getting cached/setup correctly in ::Lite.  First attempt to fix this.
  - Bumped version dependency of Ima::DBI::Contextual due to bugfixes/enhancements there.

1.031   2012-02-29
  - Special leap-day release.
  - Documented the limit/offset/order_by way of calling ->search_where

1.030   2012-02-22
  - Fixed regression in 1.029 which caused db_Main to be reset each time
    a new connection was defined.
  - UPGRADE RECOMMENDED!

1.029   2012-02-13
  - Added support for master/slaves setup.
  - Multiple slaves are supported (there can only be one master).
  - Slaves are always read from *unless* you are within a transaction.
    In that case, all reads go to the master.
  - All writes go to the master.
  - You can change your slave connection at any time via ->switch_slave($debug).
  - Passing a true value to ->switch_slave will cause a little message to be printed
    to STDERR which looks like:
      "[Debug] Switched slave from DBI:mysql:dbname:host1 to DBI:mysql:dbname:host2"
  - Also updated the 'pagination' test so that it runs quicker.

1.028   2012-01-30
  - No longer die on a failed class-load.

1.027   2012-01-10
  - Setting $pager->page_number( 1 ) now works.

1.026   2011-11-18
  - do_transaction will now look at wantarray and execute the transaction coderef
    in the appropriate list or scalar context, and return the appropriate result.

1.025   2011-11-07
  - Fixed POD for count_search_where.  (ray.baksh++)

1.024   2011-11-04
  - Calling sth_to_objects(...) would return a list with a single 'undef' element
    instead of returning nothing when no results were found.

1.023   2011-10-27
  - Now if you want to override the default 'getter' or 'setter' behavior for a
    field in your objects, you can do so by defining _set_foo( $self, $val ) or
    _get_foo( $self ).
  - Thanks Eric for the idea!  ehayes.inflection++

1.022   2011-09-19
  - POD change from app::* to App::db::* namespace.

1.021   2011-06-14
  - Upgrade Recommended!!
  - BUG FIX: `before_update` triggers which caused changes to an object did not get those
    changes saved as part of the final update.

1.020   2011-06-05
  - Updates to work with changes in the public interface of Ima::DBI::Contextual.
  - No other changes.

1.019   2011-04-08
  - Updated POD to use the new lowercase classname standard - as in .NET's Entity Framework.

1.018   2011-04-07
  - Added ->trace( 1:0 ) to enable/disable SQL logging to STDERR via Carp::cluck.

1.017   2010-12-09
  - Upgrade Recommended!!
  - BUG FIX: before_update_<field> triggers were not able to change the value of object data.
    So if you had a before_update_password trigger that hashed the password before
    the plaintext password were stored in the database, it would not work as expected.

1.016   2010-11-17
  - Included Class::DBI::Lite::Dataset - a class with helper methods useful for
    fetching a recordset useful for paged displays.

1.015   2010-09-17  ** BEHAVIOR CHANGE!!!! **
  - has_many(...) no longer adds a trigger for after_delete that would remove whatever
    the 'has_many' reference declares.  Reason? This should be handled by the database,
    not by the application.

1.014   2010-09-09
  - root() is initialized to your application's base class (eg: My::Model).
    This helps segment database handles even further.

1.013   2010-08-08
  - This release fixes that bug related to running multiple applications under
    different VirtualHosts on the same apache server.

1.012   2010-05-20
  - This release addresses several issues related to running multiple applications
    under different VirtualHosts on the same apache server.

1.011   2010-05-19
  - Removed dependency on DBIx::ContextualFetch.
  - Removed dependency on Ima::DBI - lateral move to Ima::DBI::Lite.  See Ima::DBI::Lite
    for details on why...

1.010   2010-05-13
  - My::Thing->sql_pager(...) now correctly handles {order_by => ...}

1.009   2010-04-26
  - Class::DBI::Lite::mysql was incorrectly parsing `SHOW COLUMNS FROM <table>`
    and therefore not returning the correct information when you called:
      My::Type->get_table_info->column('foo')->enum_values
    This is now fixed.

1.008   2010-04-02
  - Pager was reporting incorrect stop_item before items() or next_page() were called.
  - That caused problems when you wanted to say...
    "Showing items X - Y of Z total items"
    ...before you actually showed them.

1.007   2010-04-02
  - Pager now does the right thing when you want to page without specifying
    the field to sort by.
  - New method $pager->navigations( [$padding] ).

1.006   2010-03-27
  - Pager was not handling 'has_next' correctly - now it does.

1.005   2010-03-22
  - Renamed has_a method as belongs_to.
  - Added has_one method.
  - Added test t/010-basic/000-AAAsetup.t that creates the SQLite database structure.
  - Added the ability to pass search_where-style arguments to the has-many relationship
    methods.  So if My::Artist has_many My::Album as 'albums' you can now do:
      my @albums = $artist->albums({name => { LIKE => 'Best of%'}}, {order_by => 'name DESC'});
    * A SPECIAL thanks to Tudor Marghidanu for finding a bug and for suggesting this update.

1.004   2010-03-16
  - Had to update Changes and META.yml.

1.003   2010-03-16
  - Added Class::DBI::Lite::Pager
  - Added 2 new methods: pager(...) and sql_pager(...)

lib/Class/DBI/Lite/Pager.pm

package Class::DBI::Lite::Pager;

use strict;
use warnings 'all';


sub new
{
  my ($class, %args) = @_;
  
  my $s = bless {
    data_sql      => undef,
    count_sql     => undef,
    sql_args      => [ ],
    where         => undef,
    order_by      => undef,
    class         => undef,
    page_number   => 1,
    page_size     => 10,
    total_items   => 0,
    start_item    => 1,
    stop_item     => 0,
    %args,
    _fetched_once => 0,
  }, $class;

  ($s->{stop_item}) = sort { $a <=> $b } (
    $s->page_number * $s->page_size,
    $s->total_items
  );
  $s->{start_item} = ( $s->page_number - 1 ) * $s->page_size + 1;

  return $s;
}# end new()


# Public read-only properties:
sub page_number { shift->{page_number} }
sub page_size   { shift->{page_size} }
sub total_items { shift->{total_items} }
sub total_pages { shift->{total_pages} }
sub start_item  { shift->{start_item} }
sub stop_item   { shift->{stop_item} }
sub has_prev    { shift->{page_number} > 1 }

sub has_next
{
  my $s = shift;
  $s->{total_pages} > $s->{page_number};
}# end has_next()

sub _has_more
{
  my $s = shift;
  
  if( $s->{_fetched_once} )
  {
    $s->{page_number} < $s->{total_pages};
  }
  else
  {
    $s->{total_pages} >= $s->{page_number};
  }# end if()
}# end _has_more()

*items = \&next_page;

sub next_page
{
  my $s = shift;
  
  return unless $s->_has_more;
  
  if( $s->{_fetched_once}++ )
  {
    $s->{page_number}++;
  }# end if()

  ($s->{stop_item}) = sort { $a <=> $b } (
    $s->page_number * $s->page_size,
    $s->total_items
  );
  $s->{start_item} = ( $s->page_number - 1 ) * $s->page_size + 1;
  
  my $offset = $s->_offset;
  
  if( $s->{data_sql} )
  {
    my $limit = " LIMIT $offset, @{[ $s->{page_size} ]} ";
    my $order_by = $s->{order_by} ? " ORDER BY $s->{order_by} " : "";
    my $sth = $s->{class}->db_Main->prepare( "$s->{data_sql} $order_by $limit" );
    $sth->execute( @{ $s->{sql_args} } );
    return $s->{class}->sth_to_objects( $sth );
  }
  else
  {
    return $s->{class}->search_where(
      $s->{where},
      {
        order_by  => $s->{order_by} || undef,
        limit     => $s->page_size,
        offset    => $offset,
      }
    );
  }# end if()
}# end next_page()


sub prev_page
{
  my $s = shift;
  
  return unless $s->has_prev;
  
  $s->{page_number}-- if $s->{_fetched_once}++;
  $s->{stop_item} = $s->page_number * $s->page_size;
  $s->{stop_item} = $s->total_items if $s->stop_item > $s->total_items;
  $s->{start_item} = ( $s->page_number - 1 ) * $s->page_size + 1;
  $s->{start_item} = 0 if $s->{start_item} < 0;
  
  my $offset = $s->_offset;
  
  if( $s->{data_sql} )
  {
    my $limit = " LIMIT $offset, @{[ $s->{page_size} ]} ";
    my $sth = $s->{class}->db_Main->prepare( "$s->{data_sql} $limit" );
    $sth->execute( @{ $s->{sql_args} } );
    return $s->{class}->sth_to_objects( $sth );
  }
  else
  {
    return $s->{class}->search_where(
      $s->{where},
      {
        order_by  => $s->{order_by} || undef,
        limit     => $s->page_size,
        offset    => $offset,
      }
    );
  }# end if()
}# end prev_page()


sub navigations
{
  my ($s, $padding) = @_;
  
  $padding ||= 5;
  
  # Wiggle the start and stop out of the data we have:
  my $start = $s->page_number - $padding > 0
              ? $s->page_number - $padding
              : 1;
  my $stop  = $s->page_number + $padding <= $s->total_pages
              ? $s->page_number + $padding
              : $s->total_pages;
  
  # Now:
  if( $stop - $start < ( $padding * 2 ) + 1 )
  {
    # Need to add more pages:
    if( $start == 1 && $stop < $s->total_pages )
    {
      while( ( $stop - $start < ( $padding * 2 ) ) && $stop < $s->total_pages )
      {
        $stop++;
      }# end while()
    }
    elsif( $stop == $s->total_pages && $start > 1 )
    {
      while( ( $stop - $start < ( $padding * 2 ) ) && $start > 0 )
      {
        $start--;
      }# end while()
    }# end if()
  }# end if()
  
  return ( $start, $stop );
}# end navigations()


sub _offset
{
  my $s = shift;
  $s->{page_number} == 1 ? 0 : ($s->{page_number} - 1) * $s->{page_size};
}# end _offset()

1;# return true:

=pod

=head1 NAME

Class::DBI::Lite::Pager - Page through your records, easily.

=head1 SYNOPSIS

=head2 Paged Navigation Through Large Datasets

  # Say we're on page 1 of a list of all 'Rock' artists:
  my $pager = app::artist->pager({
    genre => 'Rock',
  }, {
    order_by    => 'name ASC',
    page_number => 1,
    page_size   => 20,
  });

  # -------- OR -----------
  my $pager = app::artist->sql_pager({
    data_sql  => "SELECT * FROM artists WHERE genre = ?",
    count_sql => "SELECT COUNT(*) FROM artists WHERE genre = ?",
    sql_args  => [ 'Rock' ],
  }, {
    page_number => 1,
    page_size   => 20,
  });
  
  # Get the first page of items from the pager:
  my @artists = $pager->items;
  
  # Is the a 'previous' page?:
  if( $pager->has_prev ) {
    print "Prev page number is " . ( $pager->page_number - 1 ) . "\n";
  }
  
  # Say where we are in the total scheme of things:
  print "Page " . $pager->page_number . " of " . $pager->total_pages . "\n";
  print "Showing items " . $pager->start_item . " through " . $pager->stop_item . " out of " . $pager->total_items . "\n";
  
  # Is there a 'next' page?:
  if( $pager->has_next ) {
    print "Next page number is " . ( $pager->page_number + 1 ) . "\n";
  }
  
  # Get the 'start' and 'stop' page numbers for a navigation strip with 
  # up to 5 pages before and after the 'current' page:
  my ($start, $stop) = $pager->navigations( 5 );
  for( $start..$stop ) {
    print "Page $_ | ";
  }

=head2 Fetch Huge Datasets in Small Chunks

  # Fetch 300,000,000 records, 100 records at a time:
  my $pager = app::citizen->pager({
    country => 'USA'
  }, {
    order_by    => 'last_name, first_name',
    page_size   => 100,
    page_number => 1,
  });
  while( my @people = $pager->next_page ) {
    # We only got 100 people, instead of swamping the 
    # database by asking for 300M records all at once:
  }

=head1 DESCRIPTION

Paging through records should be easy.  C<Class::DBI::Lite::Pager> B<makes> it easy.

=head1 CAVEAT EMPTOR

This has been tested with MySQL 5.x and SQLite.  It should work with any database
that provides some kind of C<LIMIT index, offset> construct.

To discover the total number of pages and items, 2 queries must be performed:

=over 4

=item 1 First we do a C<SELECT COUNT(*) ...> to find out how many items there are in total.

=item 2 One or more queries to get the records you've requested.

If running 2 queries is going to cause your database server to catch fire, please consider rolling your own pager
or finding some other method of doing this.

=back

=head1 CONSTRUCTOR

=head2 new( page_number => 1, page_size => 10 )

Returns a new Pager object at the page number and page size specified.

=head1 PUBLIC PROPERTIES

=head2 page_number

Read only.  Returns the page number.

=head2 page_size

Read only.  Returns the page size.

=head2 total_pages

Read only.  Returns the total number of pages in the Pager.

=head2 total_items

Read only.  Returns the total number of records in all the pages combined.

=head2 start_item

Read only.  Returns the index of the first item in this page's records.

=head2 stop_item

Read only.  Returns the index of the last item in this page's records.

=head2 has_next

Read only.  Returns true or false depending on whether there are more pages B<after> the current page.

=head2 has_prev

Read only.  Returns true or false depending on whether there are more pages B<before> the current page.

=head1 PUBLIC METHODS

=head2 items( )

Returns the next page of results.  Same as calling C<next_page()>.  Purely for syntax alone.

=head2 next_page( )

Returns the next page of results.  If called in list context, returns an array.  If 
called in scalar context, returns a L<Class::DBI::Lite::Iterator>.

If there is not a next page, returns undef.

=head2 prev_page( )

Returns the previous page of results.  If called in list context, returns an array.  If 
called in scalar context, returns a L<Class::DBI::Lite::Iterator>.

If there is not a previous page, returns undef.

=head2 navigations( [$padding = 5] )

OK - grab a cup of coffee, then come back for the explanation.

Ready?  Good.

Say you have a C<$pager>:

  my $pager = app::album->pager(undef, {
    page_size => 10,
    page_number => 1,
  });

Then you want to make your paging navigation with at least 10 pages shown, and a
maximum of 5 pages to either side of the "current" page (like Google).

  1  2  3  4  5  6  7  8  9  10 11

On the first page you I<could> just do:

  for( 1..10 ) {
    # print a link to that page.
  }

...but...when you get toward the middle or off to the end, it gets weird.

Tah-Dah!

  my ($start, $stop) = $pager->navigations( 5 );

Now you can simply do:

  for( $start..$stop ) {
    # print a link to that page:
  }

B<It> will always do the right thing - will I<you>?

So when you're on page 7 it will look like this:

  2  3  4  5  6  7  8  9  10  11  12

Then, if there were 20 pages in your entire resultset, page 20 would look like this:

  10  11  12  13  14  15  16  17  18  19  20

Great, huh?

=head1 AUTHOR

Copyright John Drago <jdrago_999@yahoo.com>.  All rights reserved.

=head1 LICENSE

This software is B<Free> software and may be used and redistributed under the
same terms as perl itself.

=cut

